Sources/eppo/RuleEvaluator.swift

```diff
@@ -3,131 +3,259 @@ import Semver
-                     .contains(a.lowercased());
+            .contains(a.lowercased());
     }
 }
 
-public class RuleEvaluator {
+struct FlagEvaluation {
+    let flagKey: String
+    let subjectKey: String
+    let subjectAttributes: SubjectAttributes
+    let allocationKey: Optional<String>
+    let variation: Optional<UFC_Variation>
+    let variationType: [UFC_VariationType]
+    let extraLogging: [String: String]
+    let doLog: Bool
+    
+    static func matchedResult(
+        flagKey: String,
+        subjectKey: String,
+        subjectAttributes: SubjectAttributes, 
+        allocationKey: Optional<String>, 
+        variation: Optional<UFC_Variation>,
+        variationType: [UFC_VariationType],
+        extraLogging: [String: String],
+        doLog: Bool
+    ) -> FlagEvaluation {
+        return FlagEvaluation(
+            flagKey: flagKey,
+            subjectKey: subjectKey,
+            subjectAttributes: subjectAttributes,
+            allocationKey: allocationKey,
+            variation: variation,
+            variationType: variationType,
+            extraLogging: extraLogging,
+            doLog: doLog
+        )
+    }
+    
+    static func noneResult(flagKey: String, subjectKey: String, subjectAttributes: SubjectAttributes) -> FlagEvaluation {
+        return FlagEvaluation(
+            flagKey: flagKey,
+            subjectKey: subjectKey,
+            subjectAttributes: subjectAttributes,
+            allocationKey: Optional<String>.none,
+            variation: Optional<UFC_Variation>.none,
+            variationType: [],
+            extraLogging: [:],
+            doLog: false
+        )
+    }
+}
+
+
+public class FlagEvaluator {
+    
+    let sharder: Sharder
+    
+    init(sharder: Sharder) {
+        self.sharder = sharder
+    }
+    
     enum Errors : Error {
         case UnexpectedValue
     }
-
-    static func findMatchingRule(
-        _ subjectAttributes: SubjectAttributes,
-        _ rules: [TargetingRule]) throws -> TargetingRule?
-    {
-        for rule in rules {
-            if try matchesRule(subjectAttributes, rule) {
-                return rule;
+    
+    func evaluateFlag(
+        flag: UFC_Flag,
+        subjectKey: String,
+        subjectAttributes: SubjectAttributes
+    ) -> FlagEvaluation {
+        if (!flag.enabled) {
+            return FlagEvaluation.noneResult(
+                flagKey: flag.key,
+                subjectKey: subjectKey,
+                subjectAttributes: subjectAttributes
+            )
+        }
+        
+        let now = Date()
+        for allocation in flag.allocations {
+            // # Skip allocations that are not active
+            if let startAt = allocation.startAt, now < startAt {
+                continue
+            }
+            if let endAt = allocation.endAt, now > endAt {
+                continue
+            }
+            
+            if matchesRules(subjectAttributes: subjectAttributes, rules: allocation.rules ?? []) {
```

Comment:
An annoying "feature" we need to continue supporting is implicitly passing the subject key as the `id` attribute if an `id` attribute is not specified. Slack thread: https://eppo-group.slack.com/archives/C06N7AA6M5F/p1714767775054549

Reply:
Adding a test for this in the sdk-test-data repo

Reply:
Do you have an example of this? Checked python and android without seeing it.